packages-exp/auth-compat-exp/src/auth.ts

```diff
@@ -0,0 +1,269 @@
+import '@firebase/installations';
+import { Observer, Unsubscribe, ErrorFn } from '@firebase/util';
+import { User } from './user';
+import {
+  convertComfirmationResult,
+  convertCredential
+} from './user_credential';
+
+export class Auth extends AuthImplCompat<User> implements compat.FirebaseAuth {
+  readonly app: FirebaseApp;
+
+  constructor(app: FirebaseApp) {
+    const { apiKey, authDomain } = app.options;
+
+    // TODO(avolkovi): Implement proper persistence fallback
+    const hierarchy = [impl.indexedDBLocalPersistence].map<persistenceImpl>(
+      _getInstance
+    );
+
+    // TODO: platform needs to be determined using heuristics
+    assert(apiKey, app.name, AuthErrorCode.INVALID_API_KEY);
+    const config: externs.Config = {
+      apiKey,
+      authDomain,
+      apiHost: DEFAULT_API_HOST,
+      tokenApiHost: DEFAULT_TOKEN_API_HOST,
+      apiScheme: DEFAULT_API_SCHEME,
+      sdkClientVersion: _getClientVersion(ClientPlatform.BROWSER)
+    };
+
+    super(app.name, config, User);
+    this.app = app;
+
+    // This promise is intended to float; auth initialization happens in the
+    // background, meanwhile the auth object may be used by the app.
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this._initializeWithPersistence(
+      hierarchy,
+      impl.browserPopupRedirectResolver
+    );
+  }
+
+  _asExtern(): externs.Auth {
+    // We need this unsafe cast before calling impl.* methods since our implementation of
+    // setPersistence, onAuthStateChanged and onIdTokenChanged has a different signature
+    return (this as unknown) as externs.Auth;
+  }
+
+  applyActionCode(code: string): Promise<void> {
+    return impl.applyActionCode(this._asExtern(), code);
+  }
+
+  checkActionCode(code: string): Promise<compat.ActionCodeInfo> {
+    return impl.checkActionCode(this._asExtern(), code);
+  }
+
+  confirmPasswordReset(code: string, newPassword: string): Promise<void> {
+    return impl.confirmPasswordReset(this._asExtern(), code, newPassword);
+  }
+
+  async createUserWithEmailAndPassword(
+    email: string,
+    password: string
+  ): Promise<compat.UserCredential> {
+    return convertCredential(
+      impl.createUserWithEmailAndPassword(this._asExtern(), email, password)
+    );
+  }
+  // currentUser: User | null;
+  fetchSignInMethodsForEmail(email: string): Promise<string[]> {
+    return impl.fetchSignInMethodsForEmail(this._asExtern(), email);
+  }
+  isSignInWithEmailLink(emailLink: string): boolean {
+    return impl.isSignInWithEmailLink(this._asExtern(), emailLink);
+  }
+  async getRedirectResult(): Promise<compat.UserCredential> {
+    const credential = await impl.getRedirectResult(
+      this._asExtern(),
+      impl.browserPopupRedirectResolver
+    );
+    if (!credential) {
+      return {
+        credential: null,
+        user: null
+      };
+    }
+    return convertCredential(Promise.resolve(credential!));
```

Comment:
`!` is unnecessary. The if statement above this guarantees credential is defined

Reply:
done